Remove activities on schedule update

alxdarksage opened the pull request:
OBSOLETE. FOR REFERENCE ONLY. THERE IS ANOTHER PR WITH THESE CHANGES AS WELL AS UPDATES BASED ON COMMENTS.

alxdarksage committed:
Delete scheduled activities when schedule plan changes. Needs more tests

alxdarksage committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/BridgePF into remove-activities-on-schedule-update

alxdarksage committed:
Added tests, index helper causing issues

alxdarksage committed:
Fixing tests

alxdarksage committed:
Some screwing around with secondary index: now only including keys, so it works with the DynamoIndexHelper

alxdarksage committed:
Removing 'ref' from schedule object, adding schedulePlanGuid to SAs, fixing up JSON

alxdarksage committed:
Updates looking at the diff

DwayneJengSage commented on app/org/sagebionetworks/bridge/dynamodb/DynamoScheduledActivity.java:
```diff
@@ -88,22 +95,26 @@ public void setTimeZone(DateTimeZone zone) {
     }
 
     @Override
+    @JsonDeserialize(using = DateTimeDeserializer.class)
```
The unit tests don't test deserializing scheduledOn or expiresOn. Can we add that to the unit tests?

alxdarksage replied:
Added, though we always use the time zone you set separately.

DwayneJengSage commented on test/org/sagebionetworks/bridge/dynamodb/DynamoScheduledActivityDaoTest.java:
```diff
@@ -130,11 +132,43 @@ public void createUpdateDeleteActivities() throws Exception {
+        activityDao.saveActivities(activities);
+
+        // Get one schedule plan GUID to delete and the initial count
+        String schedulePlanGuid = activities.get(0).getSchedulePlanGuid();
```
This is dangerous, as this schedulePlanGuid might come from an unrelated test. It would be safer to create a known schedule plan up front so we know what the schedule plan guid is, and we can target that specific schedule.

One side benefit is, if we control the schedule plan, we can use absolute counts instead of regular counts for this test.

alxdarksage replied:
Add schedule plan with unique guid and delete those related activities.

DwayneJengSage commented on app/org/sagebionetworks/bridge/dynamodb/DynamoScheduledActivityDao.java:
```diff
@@ -131,5 +139,23 @@ public void deleteActivities(String healthCode) {
+        List<ScheduledActivity> activitiesToDelete = schedulePlanIndex
+                .query(DynamoScheduledActivity.class, "schedulePlanGuid", schedulePlanGuid, null)
+                .stream()
+                .filter(act -> ScheduledActivityStatus.DELETABLE_STATUSES.contains(act.getStatus()))
```
Can we add a unit test to test that activities without deletable statuses aren't deleted?

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/SchedulePlanServiceImplMockTest.java:
```diff
@@ -0,0 +1,67 @@
+        
+        ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);
+        verify(activityService).deleteActivitiesForSchedulePlan(captor.capture());
+        assertEquals("BBB", captor.getValue());
```
Instead of using ArgumentCaptor<String>, you can just say deleteActivitiesForSchedulePlan(eq("BBB")) or even just deleteActivitiesForSchedulePlan("BBB"). (Here and 3 other examples in this PR.)

alxdarksage replied:
Changed.

alxdarksage commented:
Sorry, I redid my fork this morning and orphaned this PR. I've created another PR pulling down all these changes, with code addressing all your remarks. I'll leave this PR here so you can reference, than close it out and we'll merge the other one. Thanks.
